daemon/src/lib.rs

```diff
@@ -140,6 +148,38 @@ where
         .create(None)
         .spawn(&mut tasks);
 
+        let (endpoint_addr, endpoint_context) = Context::new(None);
+
+        let ping_address = xtra_libp2p_ping::Actor::new(endpoint_addr.clone(), PING_INTERVAL)
+            .create(None)
+            .spawn(&mut tasks);
+
+        let endpoint = Endpoint::new(
+            TokioTcpConfig::new(),
+            identity.libp2p,
+            ENDPOINT_CONNECTION_TIMEOUT,
+            [(
+                xtra_libp2p_ping::PROTOCOL_NAME,
+                xtra::message_channel::StrongMessageChannel::clone_channel(&ping_address),
+            )],
+        );
+
+        tasks.add(endpoint_context.run(endpoint));
+
+        let libp2p_socket = libp2p_socket_from_legacy_networking(&p2p_socket);
+        let endpoint_listen =
+            xtra_libp2p::create_listen_tcp_multiaddr(&libp2p_socket).expect("to parse properly");
+
+        let (supervisor, _listener_actor) = supervisor::Actor::with_policy(
+            move || {
+                let endpoint_addr = endpoint_addr.clone();
+                let endpoint_listen = endpoint_listen.clone();
+                listener::Actor::new(endpoint_addr, endpoint_listen)
```

Comment:
These also seem app-specific because they hint at a policy of "one app is either dialer or listener". `xtra-libp2p` was meant to be policy-free to make it more applicable to many usecases.

Additionally, having that code in the `xtra-libp2p` crate means each binary needs to compile it even though it only uses a part of it. For now, we could put them into the `daemon` crate but long-term, the actor systems would be better of being in the respective binary crate along-side with these use-case specific actors.

As a rule of thumb, generic code that can be built through composition should likely live in a separate crate, esp. when the goal is to release the building block as its own crate. Smaller API surface means more stable API surface :)

Reply:
I'm happy to extract `listener` and `dialer` to separate crates if that suits you.
Optionally, I could hide them behind feature flags? 
 
The trouble with policy-free crates is that they are less typically less usable, making it cumbersome to be quickly used in a project. I think the building blocks added recently, even if they hint that they are usable only for one party being dialer, the other listener, are still generic enough to be used elsewhere, and as such have nothing to do with itchysats app?

Reply:
> are still generic enough to be used elsewhere, and as such have nothing to do with itchysats app?

Yes but you don't know the requirements of "elsewhere", so really it is premature optimisation to assume that a different usecase will need the same policy.

> I'm happy to extract `listener` and `dialer` to separate crates if that suits you.

Ideally, they would sit in `maker` and `taker` but we can't do that until we move the actor systems out of `daemon` so I would put it in `daemon` for now.

> The trouble with policy-free crates is that they are less typically less usable, making it cumbersome to be quickly used in a project.

We are probably splitting hairs at this point. Policy-free code is typically _more useful_ because it applies to a wider range of usecases (keeping policy out is one of the core guidelines of the linux kernel for example [^1]). Does that make it less usable because you need to write a couple of lines of code for your usecase? Hard to say.

[^1]: As I have learnt from @tcharding, can't quote another source unfortunately :)

Reply:
I couldn't find a kernel reference for you but here is one from 'The Art of UNIX Programming', which is an awesome book in my opinion.

https://www.linuxtopia.org/online_books/programming_books/art_of_unix_programming/ch01s06_3.html

Reply:
I wonder how this copy arrived in my bookshelf!

![image](https://user-images.githubusercontent.com/9418575/165250273-60fc67dd-288a-4201-98cc-091f3cd46abd.jpeg)

Reply:
thank you for the comments, I will keep this is in mind when working in this library in the future. Really insightful.